### PR TITLE
Feature/43 type system

### PR DESCRIPTION
**Pull request by ppaulweber**
related to casm-lang/casm#43

### CODE REVIEW

File: src/cpp/type/Integer.cpp [hunk 1]
```diff
@@ -678,6 +679,18 @@ Layout* IntegerLayout::clone( void ) const
...
+std::size_t IntegerLayout::hash( void ) const
+{
+    auto tmp = std::hash< u64 >()( m_word[ 0 ] );
+
+    for( std::size_t c = 1; c < m_word.size(); c++ )
+    {
+        tmp = libstdhl::Hash::combine( tmp, m_word[ c ] );
+    }
+
+    return 0;
```
Comment (emmanuel099) [inline src/cpp/type/Integer.cpp]: This looks wrong to me.

File: src/cpp/type/Integer.cpp [hunk 2]
```diff
@@ -678,6 +679,18 @@ Layout* IntegerLayout::clone( void ) const
     return new IntegerLayout( *this );
 }
 
+std::size_t IntegerLayout::hash( void ) const
+{
+    auto tmp = std::hash< u64 >()( m_word[ 0 ] );
```
Comment (emmanuel099) [inline src/cpp/type/Integer.cpp]: `hash` or just `h` instead of `tmp` (tmp is so "unspecific")

File: src/cpp/type/Data.cpp [hunk 3]
```diff
@@ -184,6 +184,29 @@ u1 Data::defined( void ) const
     return m_trivial or m_data.ptr != 0;
 }
 
+std::size_t Data::hash( void ) const
+{
+    std::size_t hash = 0;
+    std::size_t seed
```
Comment (emmanuel099) [inline src/cpp/type/Data.cpp]: const

File: src/cpp/Memory.h [hunk 4]
```diff
@@ -83,25 +82,25 @@ namespace libstdhl
...
         template < typename T, typename... Args >
         inline std::shared_ptr< T > get( Args&&... args )
         {
             T tmp = T( std::forward< Args >( args )... );
 
-            auto cache = tmp.make_cache().find( tmp.make_hash() );
-            if( cache != tmp.make_cache().end() )
+            const auto tmp_hash = tmp.hash();
+
+            auto cache = tmp.cache().find( tmp_hash );
```
Comment (emmanuel099) [inline src/cpp/Memory.h]: const + rename `cache` to `it`, because find returns an iterator.

File: src/cpp/Memory.h [hunk 5]
```diff
@@ -83,25 +82,25 @@ namespace libstdhl
...
+        // through 'hash()'
         //
         template < typename T, typename... Args >
         inline std::shared_ptr< T > get( Args&&... args )
         {
             T tmp = T( std::forward< Args >( args )... );
 
-            auto cache = tmp.make_cache().find( tmp.make_hash() );
-            if( cache != tmp.make_cache().end() )
+            const auto tmp_hash = tmp.hash();
```
Comment (emmanuel099) [inline src/cpp/Memory.h]: nitpick: camelCase

File: src/cpp/type/Data.cpp (hunk 3 above)
Comment (ppaulweber) [reply src/cpp/type/Data.cpp]: ah you are right ... already fixed

File: src/cpp/Memory.h (hunk 4 above)
Comment (ppaulweber) [reply src/cpp/Memory.h]: ah you are right ... already fixed

File: src/cpp/Memory.h (hunk 5 above)
Comment (ppaulweber) [reply src/cpp/Memory.h]: ah you are right ... already fixed

**Pull request merged**